### ISSUE TITLE
Path of Exile: GPU hang with VK_EXT_shader_demote_to_helper_invocation and recent DXVK

### ISSUE THREAD

**Issue by edmondo** [daniel-schuermann/mesa#84]
**Describe the bug**

Path of Exile (standalone client) hangs at start with following ACO patch (bisected) and a recent DXVK version (eaa41eb7):

```
39de1211353ef103fc947a9690c30dca4bf1295f is the first bad commit
commit 39de1211353ef103fc947a9690c30dca4bf1295f
Author: Daniel Schürmann <daniel@schuermann.dev>
Date:   Mon Jul 15 19:38:43 2019 +0200

    amd/compiler: implement VK_EXT_shader_demote_to_helper_invocation
```

Following dmesg error is reported:

`kernel: [drm:amdgpu_job_timedout [amdgpu]] *ERROR* ring gfx timeout, but soft recovered`

**RenderDoc capture:**

RenderDoc caputure with LLVM:
https://drive.google.com/file/d/1KlwfkTwt371Xd3r7Td-C3yf7FgY3Pnw1/view?usp=sharing

**System information:**

 - GPU: Radeon RX470
 - ACO version: 39de1211353ef103fc947a9690c30dca4bf1295f

**Comment by daniel-schuermann** [comment]
Thx for the report! Seems, we should have done more testing before releasing this version.
In this case, the LLVM capture is of no help as LLVM doesn't support shader_demote, and thus, uses different shaders. (You can check this by opening the capture with ACO, it just renders fine.) I disabled shader_demote until we find the problem: f7b925a4029e8f94e07db8f80a7568cc35d4620c

**Comment by daniel-schuermann** [comment]
As turns out, shader_demote isn't enabled yet in DXVK (even with a recent build).
Would you please confirm/deny that disabling shader_demote fixes the issue?

**Comment by edmondo** [comment]
Yes, with the patch applied the issue is fixed.

I'm not sure what's trying to use the shader_demote (I think it shouldn't). I'll experiment later a bit and maybe with some luck I find a pattern.

**Comment by edmondo** [comment]
I'm back to 17d9490c and I can reproduce the hang by going from DXVK 1.2.3 to 1.3.

I tried this with both a wine patched with the extension VK_EXT_shader_demote_to_helper_invocation enabled and with a clean vanilla wine without the extension.

**Comment by daniel-schuermann** [comment]
Just to clarify: you can reproduce the hang with either ACO version (before + after shader_demote patch)? Then, it's probably a DXVK bug. Or is the bug only occuring on current ACO?

**Comment by edmondo** [comment]
Current ACO is good.

I can reproduce the hang only with an ACO version with VK_EXT_shader_demote_to_helper_invocation enabled and a DXVK greater or equal version 1.3.

With VK_EXT_shader_demote_to_helper_invocation disabled in ACO (or not implemented) everything is fine.

**Comment by edmondo** [comment]
> I tried this with both a wine patched with the extension VK_EXT_shader_demote_to_helper_invocation enabled and with a clean vanilla wine without the extension.

FYI the issue in this comment is unrelated to ACO and is caused by my winelib build of DXVK. A winelib build can bypass winevulkan.

**Comment by aqxa1** [comment]
This commit also breaks Persona 5 (RPCS3) with the following message:

`../src/amd/compiler/aco_insert_exec_mask.cpp:843: void
aco::{anonymous}::add_branch_code(aco::{anonymous}::exec_ctx&, aco::Block*):
Assertion (ctx.info[block->index].exec[0].second & mask_type_wqm) == 0' failed.`

Weirdly, master doesn't fix this. I have to revert the original commit (39de1211353ef103fc947a9690c30dca4bf1295f) completely.

**Comment by daniel-schuermann** [comment]
@aqxa1 Did you try just commenting out the assertion or does applying https://pastebin.com/gZjUQDnV fix the assertion?

**Comment by aqxa1** [comment]
@daniel-schuermann I didn't try commenting out the assertion but the patch fixes the problem.

**Comment by daniel-schuermann** [comment]
should be fixed with 60ef22f5a532713ac3834d0b7cefa334969ed513 in backend.
Master will receive the update probably on Monday (it works currently as demote is disabled).

**Comment by edmondo** [comment]
Thanks!
I confirm backend fixes the issue.
